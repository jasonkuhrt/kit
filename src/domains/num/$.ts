/**
 * Numeric types and utilities with branded types for mathematical constraints.
 *
 * Provides branded number types (Positive, Negative, Even, Odd, etc.) with
 * runtime validation, mathematical operations, range types, and specialized
 * numeric domains like Complex, Ratio, and BigInt. Includes type guards,
 * ordering, and equivalence utilities.
 */
export * as Num from './$$.js'
