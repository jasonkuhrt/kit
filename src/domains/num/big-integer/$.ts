export * as BigInteger from './$$.ts'
