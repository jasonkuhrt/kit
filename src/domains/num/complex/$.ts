export * as Complex from './$$.js'
