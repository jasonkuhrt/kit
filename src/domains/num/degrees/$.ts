// Export the namespace containing all Degrees operations
export * as Degrees from './$$.js'
