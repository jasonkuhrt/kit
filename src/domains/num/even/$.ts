// Export the namespace containing all Even operations
export * as Even from './$$.ts'
