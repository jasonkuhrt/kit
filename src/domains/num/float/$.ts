// Export the namespace containing all Float operations
export * as Float from './$$.js'
