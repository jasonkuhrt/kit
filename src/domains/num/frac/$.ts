// Export the namespace containing all Frac operations
export * as Frac from './$$.js'
