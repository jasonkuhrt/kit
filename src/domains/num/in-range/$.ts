// Export the namespace containing all InRange operations
export * as InRange from './$$.js'
