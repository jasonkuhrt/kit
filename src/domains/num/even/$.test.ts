import { expect, test } from 'vitest'
import { Even } from './$.js'

test('is', () => {
  // Valid cases - even integers
  expect(Even.is(0)).toBe(true)
  expect(Even.is(2)).toBe(true)
  expect(Even.is(4)).toBe(true)
  expect(Even.is(-2)).toBe(true)
  expect(Even.is(-4)).toBe(true)
  expect(Even.is(100)).toBe(true)

  // Invalid cases - odd integers
  expect(Even.is(1)).toBe(false)
  expect(Even.is(3)).toBe(false)
  expect(Even.is(-1)).toBe(false)
  expect(Even.is(-3)).toBe(false)

  // Invalid cases - non-integers
  expect(Even.is(2.5)).toBe(false)
  expect(Even.is(2.1)).toBe(false)
  expect(Even.is(Infinity)).toBe(false)
  expect(Even.is(NaN)).toBe(false)
  expect(Even.is('2')).toBe(false)
  expect(Even.is(null)).toBe(false)
})

test('from', () => {
  // Valid cases
  expect(Even.from(0)).toBe(0)
  expect(Even.from(2)).toBe(2)
  expect(Even.from(4)).toBe(4)
  expect(Even.from(-2)).toBe(-2)
  expect(Even.from(-4)).toBe(-4)
  expect(Even.from(100)).toBe(100)

  // Invalid cases - odd integers should throw
  expect(() => Even.from(1)).toThrow('Value must be even')
  expect(() => Even.from(3)).toThrow('Value must be even')
  expect(() => Even.from(-1)).toThrow('Value must be even')
  expect(() => Even.from(-3)).toThrow('Value must be even')

  // Invalid cases - non-integers should throw
  expect(() => Even.from(2.5)).toThrow('Value must be an integer')
  expect(() => Even.from(2.1)).toThrow('Value must be an integer')
  expect(() => Even.from(Infinity)).toThrow('Value must be an integer')
  expect(() => Even.from(NaN)).toThrow('Value must be an integer')
})

test('tryFrom', () => {
  // Valid cases
  expect(Even.tryFrom(0)).toBe(0)
  expect(Even.tryFrom(2)).toBe(2)
  expect(Even.tryFrom(4)).toBe(4)
  expect(Even.tryFrom(-2)).toBe(-2)
  expect(Even.tryFrom(-4)).toBe(-4)
  expect(Even.tryFrom(100)).toBe(100)

  // Invalid cases - should return null
  expect(Even.tryFrom(1)).toBe(null)
  expect(Even.tryFrom(3)).toBe(null)
  expect(Even.tryFrom(-1)).toBe(null)
  expect(Even.tryFrom(-3)).toBe(null)
  expect(Even.tryFrom(2.5)).toBe(null)
  expect(Even.tryFrom(2.1)).toBe(null)
  expect(Even.tryFrom(Infinity)).toBe(null)
  expect(Even.tryFrom(NaN)).toBe(null)
})
