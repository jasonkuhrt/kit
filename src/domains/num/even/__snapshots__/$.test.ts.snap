// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`from > from(-1) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-1
╠══════════════════════════════════════════════════╣ THEN THROWS ERROR
Error: Value must be even, got: -1
╚══════════════════════════════════════════════════╝"
`;

exports[`from > from(-2) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-2
╠══════════════════════════════════════════════════╣ THEN RETURNS NUMBER
-2
╚══════════════════════════════════════════════════╝"
`;

exports[`from > from(-3) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-3
╠══════════════════════════════════════════════════╣ THEN THROWS ERROR
Error: Value must be even, got: -3
╚══════════════════════════════════════════════════╝"
`;

exports[`from > from(-4) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-4
╠══════════════════════════════════════════════════╣ THEN RETURNS NUMBER
-4
╚══════════════════════════════════════════════════╝"
`;

exports[`from > from(0) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
0
╠══════════════════════════════════════════════════╣ THEN RETURNS NUMBER
0
╚══════════════════════════════════════════════════╝"
`;

exports[`from > from(1) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
1
╠══════════════════════════════════════════════════╣ THEN THROWS ERROR
Error: Value must be even, got: 1
╚══════════════════════════════════════════════════╝"
`;

exports[`from > from(2) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
2
╠══════════════════════════════════════════════════╣ THEN RETURNS NUMBER
2
╚══════════════════════════════════════════════════╝"
`;

exports[`from > from(2.1) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
2.1
╠══════════════════════════════════════════════════╣ THEN THROWS ERROR
Error: Value must be an integer, got: 2.1
╚══════════════════════════════════════════════════╝"
`;

exports[`from > from(2.5) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
2.5
╠══════════════════════════════════════════════════╣ THEN THROWS ERROR
Error: Value must be an integer, got: 2.5
╚══════════════════════════════════════════════════╝"
`;

exports[`from > from(3) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
3
╠══════════════════════════════════════════════════╣ THEN THROWS ERROR
Error: Value must be even, got: 3
╚══════════════════════════════════════════════════╝"
`;

exports[`from > from(4) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
4
╠══════════════════════════════════════════════════╣ THEN RETURNS NUMBER
4
╚══════════════════════════════════════════════════╝"
`;

exports[`from > from(100) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
100
╠══════════════════════════════════════════════════╣ THEN RETURNS NUMBER
100
╚══════════════════════════════════════════════════╝"
`;

exports[`from > from(Infinity) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
Infinity
╠══════════════════════════════════════════════════╣ THEN THROWS ERROR
Error: Value must be an integer, got: Infinity
╚══════════════════════════════════════════════════╝"
`;

exports[`from > from(NaN) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
NaN
╠══════════════════════════════════════════════════╣ THEN THROWS ERROR
Error: Value must be an integer, got: NaN
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is('2') 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
2
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(-1) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-1
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(-2) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-2
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
true
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(-3) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-3
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(-4) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-4
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
true
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(0) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
0
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
true
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(1) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
1
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(2) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
2
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
true
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(2.1) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
2.1
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(2.5) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
2.5
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(3) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
3
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(4) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
4
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
true
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(100) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
100
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
true
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(Infinity) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
Infinity
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(NaN) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
NaN
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
false
╚══════════════════════════════════════════════════╝"
`;

exports[`is > is(null) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
null
╠══════════════════════════════════════════════════╣ THEN RETURNS BOOLEAN
false
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(-1) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-1
╠══════════════════════════════════════════════════╣ THEN RETURNS NULL
null
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(-2) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-2
╠══════════════════════════════════════════════════╣ THEN RETURNS NUMBER
-2
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(-3) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-3
╠══════════════════════════════════════════════════╣ THEN RETURNS NULL
null
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(-4) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
-4
╠══════════════════════════════════════════════════╣ THEN RETURNS NUMBER
-4
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(0) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
0
╠══════════════════════════════════════════════════╣ THEN RETURNS NUMBER
0
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(1) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
1
╠══════════════════════════════════════════════════╣ THEN RETURNS NULL
null
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(2) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
2
╠══════════════════════════════════════════════════╣ THEN RETURNS NUMBER
2
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(2.1) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
2.1
╠══════════════════════════════════════════════════╣ THEN RETURNS NULL
null
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(2.5) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
2.5
╠══════════════════════════════════════════════════╣ THEN RETURNS NULL
null
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(3) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
3
╠══════════════════════════════════════════════════╣ THEN RETURNS NULL
null
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(4) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
4
╠══════════════════════════════════════════════════╣ THEN RETURNS NUMBER
4
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(100) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
100
╠══════════════════════════════════════════════════╣ THEN RETURNS NUMBER
100
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(Infinity) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
Infinity
╠══════════════════════════════════════════════════╣ THEN RETURNS NULL
null
╚══════════════════════════════════════════════════╝"
`;

exports[`tryFrom > tryFrom(NaN) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
NaN
╠══════════════════════════════════════════════════╣ THEN RETURNS NULL
null
╚══════════════════════════════════════════════════╝"
`;
